File: devenv~/RuntimeTypescriptTests/tests/test-basic-inheritance.ts
import { UnityInstance } from "../UnityInstance.js";
import { assertEquals, SampleValues } from "./test-utilities.js";

export function RunTests(unityInstance: UnityInstance) {

    // Get module
    const module = unityInstance.Module;

    // Test TestBasicInheritance
    const testBasicInheritanceInstance = module.static["Nahoum.UnityJSInterop.Tests"].TestBasicInheritance.GetInstance();
    assertEquals(testBasicInheritanceInstance.GetString().value, SampleValues.TestString, "TestBasicInheritance.GetString works");

    // Test SuperTestBasicInheritance
    const superTestBasicInheritanceInstance = module.static["Nahoum.UnityJSInterop.Tests"].SuperTestBasicInheritance.GetInstanceSuper();
    assertEquals(superTestBasicInheritanceInstance.GetString().value, SampleValues.TestString, "SuperTestBasicInheritance.GetString works");
    assertEquals(superTestBasicInheritanceInstance.GetAsBaseClass().GetString().value, SampleValues.TestString, "SuperTestBasicInheritance.GetAsBaseClass works");
}

File: devenv~/Assets/Tests/TestBasicInheritance.cs
namespace Nahoum.UnityJSInterop.Tests
{
    [ExposeWeb]
    public interface ITestInterfaceForBaseClass
    {
        string GetString();
    }

    public class TestBasicInheritance : ITestInterfaceForBaseClass
    {
        [ExposeWeb]
        public static TestBasicInheritance GetInstance() => new TestBasicInheritance();

        [ExposeWeb]
        public string GetString()
        {
            return SampleValues.TestString;
        }
    }

    public class SuperTestBasicInheritance : TestBasicInheritance
    {
        [ExposeWeb]
        public static SuperTestBasicInheritance GetInstanceSuper() => new SuperTestBasicInheritance();

        [ExposeWeb]
        public TestBasicInheritance GetAsBaseClass()
        {
            return this;
        }
    }
}